File: Algorithms/Recursion/ReverseLinkedList.java
class Node{
  Node next;
  int data;
  
  public Node(int data){
    this.data = data; 
  }
}



public class YourClassNameHere {
	  
public static Node reverse(Node node,Node previous){
  Node temp;
  if(node.next == null)
       temp = node;
  else 
        temp = reverse(node.next,node);
  node.next = previous;
  return temp;
}

public static void print(Node node){
  if(node == null)
    return;
  else{
   System.out.println(node.data);
   print(node.next);
  }
    
}

  
public static void main(String[] args) {
	        Node root1 = new Node(1);
	        Node root2 = new Node(2);
	        Node root3 = new Node(3);
	        root1.next = root2;
	        root2.next = root3;
	        
	        print(root1);
	        Node head = reverse(root1,null);
	        print(head);
	        
	    }
	}

File: Problems/reverseLinkedList.js
function ListNode(val){
    this.val = val;
    this.next = null;
  }

  var reverse = function(node){
      if(node.next === null)
        return node;

    var nextNode = reverse(node.next);
    node.next.next = node;
    node.next = null;
    return nextNode;
  }


  var l1 = new ListNode(1);
  var l2 = new ListNode(2);
  var l3 = new ListNode(3);
  var l4 = new ListNode(4);
  var l5 = new ListNode(5);

  l1.next = l2;
  l2.next = l3;
  l3.next = l4;
  l4.next = l5;

  var ser = reverse(l1);
  while(ser !== null){
    console.log(ser.val);
    ser = ser.next;
  }
    
